PEFT: Beyond LoRA (#3424)
huggingface/blog: peft-beyond-lora.md

---
title: "Beyond LoRA: Can you beat the most popular fine-tuning technique?"
thumbnail: /blog/assets/peft-beyond-lora/thumbnail.png
authors:
- user: BenjaminB
- user: sayakpaul
- user: hubnemo
---

# Beyond LoRA: Can you beat the most popular fine-tuning technique?

<img src="https://huggingface.co/datasets/peft-internal-testing/peft-blog-assets/resolve/main/peft-beyond-lora/lora-celebration.png" alt="Is LoRA the best PEFT technique?" width="600"/>

# When you plan to fine-tune a model in a parameter-efficient way, think beyond LoRA

> If you want to fine-tune an open model on your own data, you are probably interested in so-called parameter-efficient fine-tuning, in short *PEFT*. This term describes techniques that significantly reduce the memory requirement to fine-tune a model. Although there are dozens of these techniques, almost everyone chooses one called “LoRA”. In this blog post, we explore if LoRA is really the best choice, what tools are available to make an informed decision, and how you can benefit from extending your horizon beyond LoRA.

Reviewer: ```suggestion
If you want to fine-tune an open model on your own data, you are probably interested in so-called parameter-efficient fine-tuning, in short *PEFT*. This term describes techniques that significantly reduce the memory requirement to fine-tune a model. Although there are dozens of these techniques, almost everyone chooses one called “LoRA”. In this blog post, we explore whether LoRA is really the best choice, what tools are available to make an informed decision, and how you can benefit from extending your horizon beyond LoRA.
```

# What is PEFT and when do you need it

> There are countless open models available, but often they're not quite good enough for your use case. Prompting may help, but usually isn't enough. Rather than training a new model from scratch, you should consider fine-tuning an existing one.

Reviewer: ```suggestion
There are countless open models available, but they often aren't quite good enough for your use case. Prompting may help, but it usually isn't enough. Rather than training a new model from scratch, you should consider fine-tuning an existing one.
```

Fine-tuning, however, is memory-hungry: you generally need enough memory to fit the whole model several times over. Quantization reduces a model's memory footprint, but quantized models can't be fine-tuned directly. So a set of techniques emerged to cut the memory needed for fine-tuning, called "parameter-efficient fine-tuning", or PEFT.

> With PEFT, you can fine-tune a model using only a fraction of that memory and even fine-tune quantized models. It offers other advantages too, such as tiny checkpoint sizes, better resistance to catastrophic forgetting, and serving multiple fine-tunes from the same base model.

Reviewer: ```suggestion
With PEFT, you can fine-tune a model using only a fraction of that memory and even fine-tune quantized models. It offers other advantages, such as tiny checkpoint sizes, greater resistance to catastrophic forgetting, and the ability to serve multiple fine-tunes from the same base model.
```

> At Hugging Face, we develop the [`PEFT` library](https://github.com/huggingface/peft), which implements many PEFT techniques behind a unified API and integrates well with the ecosystem, for example [`Transformers`](https://huggingface.co/docs/transformers/main/en/peft) and [`Diffusers`](https://huggingface.co/docs/diffusers/main/en/api/loaders/peft) . It also supports [multiple quantization methods](https://huggingface.co/docs/peft/developer_guides/quantization), enabling further accessibility in parameter-efficient fine-tuning. `PEFT` provides a good starting point whether you want to fine-tune on your own data or you're researching a new PEFT method.

Reviewer: ```suggestion
At Hugging Face, we develop the [`PEFT` library](https://github.com/huggingface/peft), which implements many PEFT techniques behind a unified API and integrates well with the ecosystem, for example [`Transformers`](https://huggingface.co/docs/transformers/main/en/peft) and [`Diffusers`](https://huggingface.co/docs/diffusers/main/en/api/loaders/peft). It also supports [multiple quantization methods](https://huggingface.co/docs/peft/developer_guides/quantization), enabling further accessibility in parameter-efficient fine-tuning. `PEFT` provides a good starting point, whether you want to fine-tune on your own data or you're researching a new PEFT method.
```

Among all PEFT techniques, LoRA is by far the most popular. Here are a few estimates:

* Of a sample of 20,834 [model cards on Hugging Face Hub](https://huggingface.co/datasets/librarian-bots/model_cards_with_metadata) that mention exactly one PEFT technique, 20,509 mention LoRA (98.4%).
* We checked which PEFT techniques are popular for image generation on an external site, too. Using a sample of 10,000 checkpoints, we found 7,111 to be LoRAs. The other identified PEFT techniques are LoCon (363) and DoRA (11, arguably a LoRA variant). That means 95.0% of PEFT checkpoints are LoRAs.
* Searching for the code snippet `from peft import <PEFT CONFIG>` on GitHub ([example GH query](https://github.com/search?q=%22from+peft+import+LoraConfig%22&type=code)), 71.3% of results are for LoRA. The runners-up are LoHa (3.7%) and AdaLoRA (3.5%).

Although these estimates are not perfect, the conclusion is nonetheless that LoRA is almost certainly by far the most common PEFT technique.

> This could just mean that LoRA works best for everyone and this fact is reflected in its usage statistics. There is, however, another possibility: LoRA was one of the earlier, popular PEFT techniques. So maybe its usage became self-reinforcing: LoRA has the highest visibility, the highest number of tutorials/examples, and it has the best support in downstream packages. Thus LoRA's popularity feeds on itself.

Reviewer: ```suggestion
This could just mean that LoRA works best for everyone, and this fact is reflected in its usage statistics. There is, however, another possibility: LoRA was one of the earlier, popular PEFT techniques. So maybe its usage became self-reinforcing: LoRA has the highest visibility, the highest number of tutorials/examples, and it has the best support in downstream packages. Thus, LoRA's popularity feeds on itself.
```

<table align="center">
  <tr>
    <td align="center"><img src="https://huggingface.co/datasets/peft-internal-testing/peft-blog-assets/resolve/main/peft-beyond-lora/benchmark-highlights.png" width="900"/></td>
  </tr>
  <tr>
    <td align="center"><em>Some results from the benchmark. When it comes to test performance and memory usage, LoRA is not necessarily the best choice. Left: MetaMathQA benchmark; right: image generation benchmark. Consult this <a href="https://huggingface.co/spaces/peft-internal-testing/PEFT-method-comparison">Space</a> for the most up-to-date results.</em></td>
  </tr>
</table>

One way to interpret the results above is to think in terms of tradeoffs, for example: How well does the model perform on the test set vs how much memory is needed to train it? If a PEFT technique cannot be beaten on both of these metrics at the same time by any other technique, it is on the *Pareto Frontier*. In other words: If you want better test accuracy, you need more memory, and if you want more memory efficiency, you have to give up on accuracy.

> Let's take a closer look at the results for the LLM Math dataset benchmark. When it comes to test accuracy vs memory, we find that LoRA is indeed on the Pareto frontier. It achieves 53.2% test accuracy and requires 22.6 GB of VRAM at the peak. There are, however, other PEFT techniques on the Pareto Frontier. For instance, [BEFT](https://huggingface.co/docs/peft/main/en/package_reference/beft) achieves 32.9% test accuracy and requires only 20.2 GB of memory at max. On the other end, we have [Lily](https://huggingface.co/docs/peft/main/en/package_reference/lily) which achieves 54.9% test accuracy but requires 25.6 GB of memory. Depending on what's more important to you, you may conclude that LoRA does not present the best tradeoff for you.

Reviewer: ```suggestion
Let's take a closer look at the results for the LLM Math dataset benchmark. When it comes to test accuracy vs memory, we find that LoRA is indeed on the Pareto frontier. It achieves 53.2% test accuracy and requires 22.6 GB of VRAM at the peak. There are, however, other PEFT techniques on the Pareto Frontier. For instance, [BEFT](https://huggingface.co/docs/peft/main/en/package_reference/beft) achieves 32.9% test accuracy and requires only 20.2 GB of memory at max. On the other end, we have [Lily](https://huggingface.co/docs/peft/main/en/package_reference/lily), which achieves 54.9% test accuracy but requires 25.6 GB of memory. Depending on what's more important to you, you may conclude that LoRA does not present the best tradeoff for you.
```

Next let's take a look at the image generation benchmark. In the [Hugging Face Space](https://huggingface.co/spaces/peft-internal-testing/PEFT-method-comparison), choose “image-gen” in the “Select Task” dropdown to show the results. The goal of the task is to learn a new concept, namely a cat plushy, and generalize it to new prompts.

<table align="center">
  <tr>
    <td align="center"><img src="https://huggingface.co/datasets/peft-internal-testing/peft-blog-assets/resolve/main/peft-beyond-lora/cat-plushy-lora.png" width="400"/></td>
  </tr>
  <tr>
    <td align="center"><em>Cat plushy image created with LoRA fine-tuned on <code>FLUX.2-klein-base-4B</code>.</em></td>
  </tr>
</table>

> For this task, the main metric is “dino similarity”, which measures how much a generated image resembles the picture from a hold out test dataset, with higher values being better. As always, we also want to keep an eye on memory usage. When plotting the Pareto Frontier of these two metrics, we find that LoRA is below that frontier. Let's get concrete numbers: LoRA achieves a similarity score of 0.697 whereas [OFT](https://huggingface.co/docs/peft/package_reference/oft) achieves 0.708; in terms of memory, LoRA requires 9.97 GB and OFT requires 9.01 GB. Therefore, OFT strictly dominates LoRA on these metrics.

Reviewer: ```suggestion
For this task, the main metric is “dino similarity”, which measures how much a generated image resembles the picture from a holdout test dataset, with higher values being better. As always, we also want to keep an eye on memory usage. When plotting the Pareto Frontier of these two metrics, we find that LoRA is below that frontier. Let's get concrete numbers: LoRA achieves a similarity score of 0.697 whereas [OFT](https://huggingface.co/docs/peft/package_reference/oft) achieves 0.708; in terms of memory, LoRA requires 9.97 GB, and OFT requires 9.01 GB. Therefore, OFT strictly dominates LoRA on these metrics.
```

# Limitations

> Objection: But the benchmarks favor one method over another!

One criticism that could be leveled at the `PEFT` benchmarks is that the choice of hyper-parameters may favor one technique over another. This is true, doing an exhaustive and fair hyper-parameter sweep with this many techniques is difficult. It is, however, very easy for everyone to contribute their own experiments to `PEFT`: If you believe that a specific PEFT technique can be improved by choosing different hyper-parameters, create a PR! We added [instructions on how to do that](https://github.com/huggingface/peft/tree/main/method_comparison#creating-new-experiments). In a similar vein, if you want to contribute a completely new benchmark, reach out to us to discuss your idea.

> Another problem with the benchmarks is that they may not fully reflect the capabilities of a specific PEFT technique. We allow to compare the techniques along many different dimensions and discover the best ones according to these tradeoffs. But it's impossible to capture all facets this way. For instance, one PEFT technique called Cartridges ([https://huggingface.co/docs/peft/package_reference/cartridges](https://huggingface.co/docs/peft/package_reference/cartridges)) was developed to compress long prompts, which is not measured in the benchmarks. Other factors can also influence the choice, for instance:

Reviewer: ```suggestion
Another problem with the benchmarks is that they may not fully reflect the capabilities of a specific PEFT technique.  We allow for comparing the techniques along many different dimensions and discover the best ones according to these tradeoffs. But it's impossible to capture all facets this way. For instance, one PEFT technique called Cartridges ([https://huggingface.co/docs/peft/package_reference/cartridges](https://huggingface.co/docs/peft/package_reference/cartridges)) was developed to compress long prompts, which is not measured in the benchmarks. Other factors can also influence the choice, for instance:
```

---
title: "Beyond LoRA: Can you beat the most popular fine-tuning technique?"
thumbnail: /blog/assets/peft-beyond-lora/thumbnail.png
authors:
- user: BenjaminB
- user: sayakpaul
- user: hubnemo
- user: kashif
---

# Beyond LoRA: Can you beat the most popular fine-tuning technique?

> <img src="https://huggingface.co/datasets/peft-internal-testing/peft-blog-assets/resolve/main/peft-beyond-lora/lora-celebration.png" alt="Is LoRA the best PEFT technique?" width="600"/>

Reviewer 2: ```suggestion
<p align="center">
    <img src="https://huggingface.co/datasets/peft-internal-testing/peft-blog-assets/resolve/main/peft-beyond-lora/lora-celebration.png" alt="Is LoRA the best PEFT technique?" width="600"/>
</p>
```

Another problem with the benchmarks is that they may not fully reflect the capabilities of a specific PEFT technique. We make it possible to compare the techniques along many different dimensions and discover the best ones according to these tradeoffs. But it's impossible to capture all facets this way. For instance, one PEFT technique called Cartridges ([https://huggingface.co/docs/peft/package_reference/cartridges](https://huggingface.co/docs/peft/package_reference/cartridges)) was developed to compress long prompts, which is not measured in the benchmarks. Other factors can also influence the choice, for instance:

* Depending on the PEFT technique, only certain layer types can be modified.
* Not all PEFT techniques support quantized base models (but we actively expand the support in `PEFT`).
* Some PEFT techniques allow [merging of the adapter](https://huggingface.co/docs/peft/main/en/developer_guides/model_merging) to reduce runtime overhead but others don't.

The benchmarks cannot fully lift the responsibility to do your research, but they can be reasonable pointers.

<table align="center">
  <tr>

>     <td align="center"><a href="https://huggingface.co/spaces/peft-internal-testing/PEFT-shop"><img src="https://huggingface.co/datasets/peft-internal-testing/peft-blog-assets/resolve/main/peft-beyond-lora/peft-shop.png" width="600"/></a></td>

Reviewer 2: ```suggestion
    <td align="center"><a href="https://huggingface.co/spaces/peft-internal-testing/PEFT-shop"><img src="https://huggingface.co/datasets/peft-internal-testing/peft-blog-assets/resolve/main/peft-beyond-lora/peft-shop.png" width="100%"/></a></td>
```
